File: platforms/bukkit/runtime/src/main/java/ru/privatenull/pnlibrary/remote/bukkit/RemoteSourceCompiler.java
package ru.privatenull.pnlibrary.remote.bukkit;

import javax.tools.JavaCompiler;
import javax.tools.ToolProvider;
import java.io.IOException;
import java.net.URL;
import java.net.URLClassLoader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class RemoteSourceCompiler {
    private static final Pattern PACKAGE = Pattern.compile("\\bpackage\\s+([A-Za-z_$][\\w$]*(?:\\.[A-Za-z_$][\\w$]*)*)\\s*;");
    private static final Pattern TYPE = Pattern.compile("\\b(?:public\\s+)?(?:final\\s+|abstract\\s+)?class\\s+([A-Za-z_$][\\w$]*)");

    static RemoteClassLoader compile(byte[] source, ClassLoader parent) throws Exception {
        String text = new String(source, StandardCharsets.UTF_8);
        Matcher type = TYPE.matcher(text);
        if (!type.find()) throw new IllegalArgumentException("remote .java file does not contain a class");
        Matcher pkg = PACKAGE.matcher(text);
        String className = (pkg.find() ? pkg.group(1) + "." : "") + type.group(1);
        Path root = Files.createTempDirectory("pnlibrary-remote-source-");
        Path file = root.resolve(className.replace('.', '/') + ".java");
        Files.createDirectories(file.getParent()); Files.write(file, source);
        JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
        if (compiler == null) throw new IllegalStateException("JavaCompiler is unavailable; run the server with a JDK");
        String classpath = System.getProperty("java.class.path", "")
                + java.io.File.pathSeparator + location(RemoteCheck.class)
                + java.io.File.pathSeparator + location(org.bukkit.plugin.java.JavaPlugin.class);
        int result = compiler.run(null, null, null, "-source", "8", "-target", "8",
                "-classpath", classpath, "-d", root.toString(), file.toString());
        if (result != 0) throw new IllegalArgumentException("remote Java policy compilation failed");
        URLClassLoader loader = new URLClassLoader(new URL[] { root.toUri().toURL() }, parent);
        return RemoteClassLoader.fromCompiled(loader, root, className);
    }

    private static String location(Class<?> type) {
        try { return new java.io.File(type.getProtectionDomain().getCodeSource().getLocation().toURI()).getPath(); }
        catch (Exception ignored) { return ""; }
    }
}

File: platforms/bukkit/runtime/src/main/kotlin/ru/privatenull/pnlibrary/remote/bukkit/RemoteSourceCompiler.kt
package ru.privatenull.pnlibrary.remote.bukkit

import org.bukkit.plugin.java.JavaPlugin
import java.io.File
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import javax.tools.ToolProvider

internal object RemoteSourceCompiler {
    private val packagePattern = Regex("\\bpackage\\s+([A-Za-z_\$][\\w\$]*(?:\\.[A-Za-z_\$][\\w\$]*)*)\\s*;")
    private val classPattern = Regex("\\b(?:public\\s+)?(?:final\\s+|abstract\\s+)?class\\s+([A-Za-z_\$][\\w\$]*)")

    fun compile(source: ByteArray, parent: ClassLoader): RemoteClassLoader {
        val text = source.toString(StandardCharsets.UTF_8)
        val simpleName = classPattern.find(text)?.groupValues?.get(1)
            ?: throw IllegalArgumentException("remote .java file does not contain a class")
        val packageName = packagePattern.find(text)?.groupValues?.get(1)
        val name = if (packageName == null) simpleName else "$packageName.$simpleName"
        val root = Files.createTempDirectory("pnlibrary-remote-source-")
        try {
            val file = root.resolve(name.replace('.', '/') + ".java")
            Files.createDirectories(file.parent)
            Files.write(file, source)
            val compiler = ToolProvider.getSystemJavaCompiler()
                ?: throw IllegalStateException("JavaCompiler is unavailable; run the server with a JDK")
            val classpath = listOf(System.getProperty("java.class.path", ""), location(RemoteCheck::class.java), location(JavaPlugin::class.java))
                .filter { it.isNotBlank() }.joinToString(File.pathSeparator)
            val result = compiler.run(null, null, null, "-source", "8", "-target", "8", "-classpath", classpath,
                "-d", root.toString(), file.toString())
            require(result == 0) { "remote Java policy compilation failed" }
            return RemoteClassLoader.compiled(root, name, parent)
        } catch (error: Throwable) {
            Files.walk(root).use { paths -> paths.sorted(Comparator.reverseOrder()).forEach { Files.deleteIfExists(it) } }
            throw error
        }
    }

    private fun location(type: Class<*>): String = try {
        File(type.protectionDomain.codeSource.location.toURI()).path
    } catch (_: Exception) { "" }
}
